flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointStatsTracker.java

```diff
@@ -227,11 +257,23 @@ void reportFailedCheckpoint(FailedCheckpointStats failed) {
             history.replacePendingCheckpointById(failed);
 
             dirty = true;
+            logCheckpointStatistics(failed);
         } finally {
             statsReadWriteLock.unlock();
         }
     }
 
+    private void logCheckpointStatistics(AbstractCheckpointStats checkpointStats) {
+        try {
+            StringWriter sw = new StringWriter();
+            MAPPER.writeValue(sw, CheckpointStatistics.generateCheckpointStatistics(checkpointStats, true));
+            String jsonDump = sw.toString();
```

Comment:
Should we save some resources by guarding this with isDebugEnabled?

Reply:
I was thinking about that, decided against complicating this line, as this method is called very rarely. But in the end sure, why not.